### ISSUE TITLE
Race condition with concurrent operations

### ISSUE THREAD

**Issue by metadaddy** [Void-ux/aiob2#2]
If you create an `aiob2.Client` instance, then use it to fire off multiple concurrent operations, multiple coroutines can be in the section between testing and setting the underlying `HTTPClient._authorization_token`, resulting in multiple authentication calls.

Example code:

```
import os
import asyncio

from aiob2 import Client

keys = [
    "some-big-file",
    "some-small-file"
]


async def download_file(client, bucket, key):
    print(f"About to download {bucket}/{key}")
    f = await client.download_file_by_name(
        file_name=key,
        bucket_name=bucket,
    )
    print(f"Downloaded {bucket}/{key}: {f.content_length} bytes")
    return f.content_length


async def main():
    print(f"Downloading objects {keys}")
    async with Client(os.environ["B2_APPLICATION_KEY_ID"], os.environ["B2_APPLICATION_KEY"]) as client:
        sizes = await asyncio.gather(*(download_file(client, "my-bucket", key) for key in keys))
    print(f"Object sizes: {sizes}")


if __name__ == '__main__':
    asyncio.run(main())
```

Output:

```
Downloading objects ['some-big-file', 'some-small-file']
About to download my-bucket/some-big-file
2023-05-04 13:17:15 INFO     aiob2.http Authenticating using static application key
About to download my-bucket/some-small-file
2023-05-04 13:17:15 INFO     aiob2.http Authenticating using static application key
Downloaded my-bucket/some-small-file: 6 bytes
Downloaded my-bucket/some-big-file: 15729152 bytes
Object sizes: ['15729152', '6']
```

The second download doesn't "see" the token retrieved by the first download.

Assuming the goal is to have the `HTTPClient` authenticate just once, I'm not sure it's possible to do 'lazy' authentication without synchronizing on `HTTPClient._authorization_token`. It might be easier to just have `HTTPClient` authenticate in its constructor. In case of authentication failure, that's probably a better (more obvious to the developer) place to throw the error than the first time the client is used.

**Comment by Void-ux** [comment]
Ah, that was an oversight. Fixed it in e7a1a18 by adding a mutex to HTTPClient.

Having an async constructor is possible, but not ideal, especially since aiob2 is designed to handle token-refreshment. Backblaze refreshes these every 24h, and its implied that they may be de-authorised before this expiration. aiob2 will refresh these tokens before sending a request that'll definitely return a 401, and refresh them if any fixable error (through refreshment) occurs.

I added a mutex into the HTTPClient.request method, which'll let the first authorization through, and delay the others while it completes, and then use the one that the first call obtained allowing them all to continue in concurrency. The new output shows only one call being made now:
```
Downloading objects ['pfp.png', 'pfp.png']
About to download aiob2-testing/pfp.png
2023-05-11 22:05:06 INFO     aiob2.http Authenticating using static application key
About to download aiob2-testing/pfp.png
Downloaded aiob2-testing/pfp.png: 33378 bytes
Downloaded aiob2-testing/pfp.png: 33378 bytes
Object sizes: ['33378', '33378']
```
